main/src/library/List.flix

```diff
@@ -968,6 +968,51 @@ namespace List {
         case Some(a) => unfoldWithIterHelper(next, xs -> k(a :: xs))
     }
 
+
+    ///
+    /// Build a list by applying the function `next` to `()`. `next` is expected to encapsulate
+    /// a stateful resource such as a file handle that can be iterated.
+    ///
+    /// `next` should return `Ok(Some(a)` to signal a new list element `Ok(a)`.
+    ///
+    /// `next` should return `Err(e)` to signal a new list element `Err(e)` and the end of building the list.
+    ///
+    /// `next` should return `Ok(None)` to signal the end of building the list
+    ///
+    pub def unfoldWithOkIter(next: () ~> Result[Option[a], e]): List[Result[a, e]] & Impure = unfoldWithOkIterHelper(next, xs -> xs)
+
+    def unfoldWithOkIterHelper(next: () ~> Result[Option[a], e], k: List[Result[a, e]] -> List[Result[a, e]]): List[Result[a, e]] & Impure = match next() {
+        case Ok(None) => k(Nil)
+        case Err(e) => k(Err(e) :: Nil)
+        case Ok(Some(a)) => unfoldWithOkIterHelper(next, xs -> k(Ok(a) :: xs))
+    }
+
+    ///
+    /// Sequences the `Result`s in `xs` and collects them
+    ///
+    pub def sequenceResult(xs: List[Result[a, e]]): Result[List[a], e] =
```

Comment:
This function signature is actually a great way to learn something:

You take a list of result and you claim to convert that to a Result[List[a], e]. 

But what if there are multiple errors? The signature tells us that you silently discard them. This is not acceptable.

A better signature is: `Result[List[a], List[e]]`. An even better signature is `Result[List[a], Nel[e]]`. Do you see why?

Once you have grasped that then look at the signature of `Validation`.

OK; with that understanding in place, and with my above comment about `unfoldWithOkIterHelper` I think these two helper functions `sequenceResult` and `sequenceOption` go away.

Reply:
(Cool how I could learn all of this just from the signature, huh?!)

Reply:
I simply wasn't aware that `Validation` actually implemented `traverse` and `sequence`. It is, however, missing a conversion from `Result` to make it more ergonomic to use. I can add that function too, if it's needed.

Is there are reason for using `Result` in some places and `Validation` in others? It seems to me that `Validation` can do everything `Result` can, but has more functionality when it comes to composition of error-cases. And implementing the monad (and functor) functionality looks entirely trivial to me.

Reply:
Validation is not a monad. Its only an applicative. (Fun exercise: Figure out why).

As for Result vs. Validation, we prefer Result because it is rare that library functions need to *accumulate* errors. 

You can add conversions to Validation, but I would prefer to avoid the use of Validation entirely and instead simply fix the type signature of unfoldWithOkIter then I think all of these issues go away.

Reply:
It is good that you are looking into these things :) but in this case I think the heavy machinery of sequence/traverse/Validation is not necessary.

Reply:
That makes sense, and I now see why `Validation` isn't a monad. What do you think should be done at the point of usage of `unfoldWithOkIter`? Because the error value of a Result [colors the function](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/). In my specific use-case I'd be in an `IOError` "context" and suddenly have to deal with a `List[IOError]` which in turn means I'd either have to "paint" the calling function, create an `Accumulated` variant of `IOError`, or in some way swallow the list into one of the already-present `IOError` variants.

The first option is generally equivalent to `Validation`, which you already said isn't wanted (I agree). Nesting an error-type inside of itself also seems to me like a very wrong way to do things, and that leaves only one option. But which error represents any error happening in multiple calls? Just an `IOError.Generic` defeats the purpose of granular errors entirely.

Reply:
I thought that `unfoldWithOkIter` stops immediately when it encounters and error, hence there is only one errors, hence Result is usable?

Reply:
No, you're right. I just got your suggestions for the above `sequence` functions and `unfoldWithOkIter` tangled up.